### ISSUE TITLE
Is there a reason why $model->delete() throws Exception?

### ISSUE THREAD

**Issue by phillip-elm** [laravel/ideas#1226]
I use PHPStorm as my primary PHP development IDE. When I'm working with models and I want to delete something, I have to catch an `Exception`, otherwise I get a warning. This warning is good - I should be catching exceptions that are thrown when things go wrong.

I'm finding that all of my model delete calls end up looking like this:

```php
$foo = Foo::first();

try {
  $foo->delete();
} catch (\Exception $e) {
  // sadness
}
```

I've briefly looked through the method, and so far the only case where an exception is thrown seems to be when the model doesn't have a primary key defined.

_Model.php:_

```php
if (is_null($this->getKeyName())) {
    throw new Exception('No primary key defined on model.');
}
```

Is there any other case where this method will throw that exception? If not, this seems a lot more like an exception thrown in development when you make a mistake, which I'd argue doesn't justify wrapping every call with `try/catch`.

I do think that an exception should be thrown in this case - but perhaps `\RuntimeException` or something that extends it instead? That way it gets thrown when `getKeyName() === null`, but it doesn't require so much boilerplate.

**Comment by sisve** [comment]
> Is there any other case where this method will throw that exception?

1. Any database errors.
2. Any exceptions from event listeners.
3. Bad luck.

**Comment by phillip-elm** [comment]
So, on that logic - why don't other methods that interact with the database or fire events throw as well? 

I'm happy to keep wrapping with `try/catch` if required, I'm more curious at this point than anything.

**Comment by drbyte** [comment]
Maybe I'm not seeing this from the right angle, but why limit it to just the `delete()`?

Couldn't you push the exception trigger up a level by changing:

```diff
-$foo = Foo::first();
+$foo = Foo::findOrFail($id); // throws a NotFound Exception if anything goes wrong in finding the record
```

At least that way you'd be sent off into an exception handler long before trying to do the delete.

**Comment by sisve** [comment]
@phillip-elm They do. You're using some plugin that checks the `@throws` annotation, and as you've found, the framework doesn't really use them. Sure, it's on the `delete()` method because the method itself throws an exception, but it doesn't care about when called code throws exceptions. And this part will very quickly explode into a huge amount of `@throws` annotations everywhere, since every callback, every event, every database call can throw an exception.

You don't have to try/catch anything. You can reduce the severity of the annotation checker you're using, or turn it off completely. Look through your tests and your logs to find the exceptions that are actually happening, and handle those.

**Comment by phillip-elm** [comment]
@sisve - I see your point - but I'm not sure the vast majority of developers using Laravel and PHPStorm in tandem would agree that turning off uncaught exception warnings is the appropriate way to go about doing this.

I'm also not saying that it's unreasonable to expect that code down the line will throw something, especially if the event listeners that can fire throw something.

What I think needs consideration is that the `delete()` function itself throws the base `Exception` class and reflects this in its docblock, which _should_ normally be caught if thrown. Turning off warnings altogether to get around this really isn't an option, especially when the exception is only thrown when a developer makes a mistake that would never find it's way to prod. Companies with any sort of code standards policy would also frown on this.

You're absolutely right - this exception should not be caught as the developer would see the uncaught exception in their development environment and fix the problem from there. That's why I believe the realistic fix here is to switch to `RuntimeException`, which typically does not cause warnings in IDEs. 

People who make the mistake of not having a PK on a model will get their error during development, and engineers who like being warned about possibly missed exceptions also continue to be warned when it's appropriate.

tl;dr:

- This change would have no affect on the intended outcome of the original exception
- The change would also not affect existing codebases that attempt to catch the original exception, as `RuntimeException extends Exception`
- Developers using IDEs would not have to meddle with settings to catch a debug exception
- Everybody's a winner

**Comment by phillip-elm** [comment]
@drbyte - I see where you're coming from, but that's not the point I'm making.

The `$foo = Foo::first();` bit was just for example. There are often times segments in code (service class patterns, for example) where you're handed a model that has already been loaded elsewhere.

**Comment by drbyte** [comment]
@phillip-elm Understood. Admittedly I took your example very literally. 
As you say, many of these things can be handled by the app, but there are some framework considerations that "could" make some specific dev-environments (specific to tooling choices like PhpStorm (of which I'm a fan too), etc) "nicer". 

I have no idea how a PR would be received about this, even if there's no BC risk, unless the person reviewing the PR is also running into the same issues exposed by their tools.

**Comment by mfn** [comment]
> switch to RuntimeException

I'd support that.

It's more a philosophical thing at some point, but basically `Exception` should almost newer be thrown (I wish it were `abstract`); we've tons of other including your `RuntimeException` example.

Good news: the usage of `Exception` is in the minority, quick count in my local installation of Laravel has shown be 14 matches whereas for `new RuntimeException` I get 30 matches.
